tests/integration/test_client_server.py

```diff
@@ -109,3 +110,24 @@ def test_endpoint_not_found(mock_server):
             endpoint='fooo',
             hostport=mock_server.hostport,
         )
+
+
+@pytest.mark.gen_test
+def test_connection_close(mock_server):
+    tchannel = TChannel(name='test')
+
+    class TestHook(EventHook):
+        def before_send_request(self, request):
+            # close the connection
+            peer = tchannel._dep_tchannel.peers.get(mock_server.hostport)
+            peer.close()
+            peer.connections[0].closed = True
+
+    tchannel.hooks.register(TestHook())
```

Comment:
the way iostream calls the callback method is 
```
    def _run_callback(self, callback, *args):
       ....
        with stack_context.NullContext():
            self._pending_callbacks += 1
            self.io_loop.add_callback(wrapper)
```
from the last line, we see the callback won't be called immediately, there is some delay. That's why even if I stop the server, the call back is triggered later.